confluent_kafka/consumer.py

```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 Confluent Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from copy import deepcopy
+
+from .cimpl import Consumer as _cConsumer
+from .serialization import SerializerError, ValueSerializerError, KeySerializerError
+
+
+class Consumer(_cConsumer):
+    """
+    .. py:function:: Consumer(config)
+
+    Producer factory
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if 'key_serializer' in kwargs or 'value_serializer' in kwargs:
+            return super(Consumer, cls).__new__(SerializingConsumer,
+                                                *args, **kwargs)
+
+        return super(Consumer, cls).__new__(cls, *args, **kwargs)
+
+
+class SerializingConsumer(Consumer):
+    __slots__ = ['_key_serializer', '_value_serializer']
+
+    def __new__(cls, conf, key_serializer=None, value_serializer=None,
+                error_cb=None, log_cb=None, stats_cb=None, throttle_cb=None):
+        raise TypeError("SerializingConsumer is a non user-instantiable class")
+
+    def __init__(self, conf, key_serializer=None, value_serializer=None,
+                 error_cb=None, log_cb=None, stats_cb=None, throttle_cb=None):
+        """
+
+        :param conf:
+        :param key_serializer:
+        :param value_serializer:
+        """
+
+        self._key_serializer = key_serializer
+        self._value_serializer = value_serializer
+
+        consumer_config = deepcopy(conf)
+        consumer_config.update({
+            'error_cb': error_cb,
+            'log_cb': log_cb,
+            'throttle_cb': throttle_cb,
+            'stats_cb': stats_cb
+        })
+
+        super(SerializingConsumer, self).__init__(consumer_config)
+
+    def poll(self, timeout=-1):
+        """
+
+        :param timeout:
+        :return:
+        """
+        msg = super(SerializingConsumer, self).poll(timeout)
+
+        if msg is None:
+            return None
+
+        if not msg.error():
+            try:
+                if self._value_serializer:
+                    msg.set_value(self._value_serializer.deserialize(msg.value(), None))
+            except SerializerError as se:
+                raise ValueSerializerError(se.message)
```

Comment:
This actually comes from the existing avro implementation I just moved it. Since we are leaving the avro subpackage untouched moving forward will create more appropriate error types. 

https://github.com/confluentinc/confluent-kafka-python/blob/master/confluent_kafka/avro/serializer/__init__.py#L35